nifi-nar-bundles/nifi-enrich-bundle/nifi-enrich-processors/src/main/java/org/apache/nifi/processors/enrich/QueryDNS.java

```diff
@@ -0,0 +1,269 @@
+@InputRequirement(InputRequirement.Requirement.INPUT_REQUIRED)
+@CapabilityDescription("A powerful DNS query processor primary designed to enrich DataFlows with DNS based APIs " +
+        "(e.g. RBLs, ShadowServer's ASN lookup) but that can be also used to perform regular DNS lookups.")
+@WritesAttributes({
+        @WritesAttribute(attribute = "enrich.dns.record*.group*", description = "The captured fields of the DNS query response for each of the records received"),
+})
+public class QueryDNS extends AbstractEnrichProcessor {
+
+    public static final PropertyDescriptor DNS_QUERY_TYPE = new PropertyDescriptor.Builder()
+            .name("DNS Query Type")
+            .description("The DNS query type to be used by the processor (e.g. TXT, A)")
+            .required(true)
+            .defaultValue("TXT")
+            .addValidator(StandardValidators.NON_EMPTY_VALIDATOR)
+            .build();
+
+    public static final PropertyDescriptor DNS_SERVER = new PropertyDescriptor.Builder()
+            .name("DNS Servers")
+            .description("A comma separated list of  DNS servers to be used. (Defaults to system wide if none is used)")
+            .required(false)
+            .addValidator(StandardValidators.NON_EMPTY_VALIDATOR)
+            .build();
+
+    public static final PropertyDescriptor DNS_TIMEOUT = new PropertyDescriptor.Builder()
+            .name("DNS Query Timeout")
+            .description("The amount of milliseconds to wait until considering a query as failed")
+            .required(true)
+            .defaultValue("1500")
+            .addValidator(StandardValidators.NON_EMPTY_VALIDATOR)
+            .build();
+
+    public static final PropertyDescriptor DNS_RETRIES = new PropertyDescriptor.Builder()
+            .name("DNS Query Retries")
+            .description("The number of attempts before giving up and moving on")
+            .required(true)
+            .defaultValue("1")
+            .addValidator(StandardValidators.NON_EMPTY_VALIDATOR)
+            .build();
+
+
+    private final static List<PropertyDescriptor> propertyDescriptors;
+    private final static Set<Relationship> relationships;
+
+    private DirContext ictx;
+
+    // Assign the default and generally used contextFactory value
+    private String contextFactory = com.sun.jndi.dns.DnsContextFactory.class.getName();;
+
+    static {
+        List<PropertyDescriptor> props = new ArrayList<>();
+        props.add(QUERY_INPUT);
+        props.add(QUERY_PARSER);
+        props.add(QUERY_PARSER_INPUT);
+        props.add(DNS_RETRIES);
+        props.add(DNS_TIMEOUT);
+        props.add(DNS_SERVER);
+        props.add(DNS_QUERY_TYPE);
+        propertyDescriptors = Collections.unmodifiableList(props);
+
+        Set<Relationship> rels = new HashSet<>();
+        rels.add(REL_FOUND);
+        rels.add(REL_NOT_FOUND);
+        relationships = Collections.unmodifiableSet(rels);
+    }
+
+    private AtomicBoolean initialized;
```

Comment:
It needs to be initialized otherwise I get:
````
ava.lang.RuntimeException: Failed while executing one of processor's OnScheduled task.
	at org.apache.nifi.controller.StandardProcessorNode.invokeTaskAsCancelableFuture(StandardProcessorNode.java:1418) ~[na:na]
	at org.apache.nifi.controller.StandardProcessorNode.access$100(StandardProcessorNode.java:93) ~[na:na]
	at org.apache.nifi.controller.StandardProcessorNode$1.run(StandardProcessorNode.java:1256) ~[na:na]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) [na:1.8.0_77]
	at java.util.concurrent.FutureTask.run(FutureTask.java:266) [na:1.8.0_77]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180) [na:1.8.0_77]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293) [na:1.8.0_77]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) [na:1.8.0_77]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) [na:1.8.0_77]
	at java.lang.Thread.run(Thread.java:745) [na:1.8.0_77]
Caused by: java.util.concurrent.ExecutionException: java.lang.reflect.InvocationTargetException
	at java.util.concurrent.FutureTask.report(FutureTask.java:122) [na:1.8.0_77]
	at java.util.concurrent.FutureTask.get(FutureTask.java:206) [na:1.8.0_77]
	at org.apache.nifi.controller.StandardProcessorNode.invokeTaskAsCancelableFuture(StandardProcessorNode.java:1401) ~[na:na]
	... 9 common frames omitted
Caused by: java.lang.reflect.InvocationTargetException: null
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[na:1.8.0_77]
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[na:1.8.0_77]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:1.8.0_77]
	at java.lang.reflect.Method.invoke(Method.java:498) ~[na:1.8.0_77]
	at org.apache.nifi.util.ReflectionUtils.invokeMethodsWithAnnotations(ReflectionUtils.java:137) ~[na:na]
	at org.apache.nifi.util.ReflectionUtils.invokeMethodsWithAnnotations(ReflectionUtils.java:125) ~[na:na]
	at org.apache.nifi.util.ReflectionUtils.invokeMethodsWithAnnotations(ReflectionUtils.java:70) ~[na:na]
	at org.apache.nifi.controller.StandardProcessorNode$1$1.call(StandardProcessorNode.java:1260) ~[na:na]
	at org.apache.nifi.controller.StandardProcessorNode$1$1.call(StandardProcessorNode.java:1256) ~[na:na]
	... 6 common frames omitted
Caused by: java.lang.NullPointerException: null
	at org.apache.nifi.processors.enrich.QueryDNS.onScheduled(QueryDNS.java:209) ~[na:na]
	... 15 common frames omitted
````